File: Intermediate/32.Solve the Topological Sorting Problem for a directed acyclic graph (DAG)/Solution.py
def dfs(v, graph, visited, stack):
    """Recursive DFS that marks visited nodes and appends vertices to stack."""
    visited[v] = True
    for neighbor in graph[v]:
        if not visited[neighbor]:
            dfs(neighbor, graph, visited, stack)
    stack.append(v)

def topological_sort(graph, vertices):
    """Returns a list of vertices in topologically sorted order."""
    visited = [False] * vertices
    stack = []
    for i in range(vertices):
        if not visited[i]:
            dfs(i, graph, visited, stack)
    # Reverse stack to get the correct topological order.
    return stack[::-1]

def main():
    vertices = int(input("Enter the number of vertices: "))
    edges = int(input("Enter the number of edges: "))
    
    # Initialize the graph as a list of lists
    graph = [[] for _ in range(vertices)]
    print("Enter each edge (source destination) separated by space:")
    for _ in range(edges):
        u, v = map(int, input().split())
        graph[u].append(v)  # Directed edge from u to v
    
    order = topological_sort(graph, vertices)
    print("Topological Order:", " ".join(map(str, order)))

if __name__ == "__main__":
    main()

File: Intermediate/32.Solve the Topological Sorting Problem for a directed acyclic graph (DAG)/Solution.cpp
#include <iostream>
#include <vector>
#include <stack>
using namespace std;

// Recursive DFS function to process a vertex and its neighbors
void dfs(int v, const vector<vector<int>> &graph, vector<bool> &visited, stack<int> &order) {
    visited[v] = true;
    for (int neighbor : graph[v]) {
        if (!visited[neighbor]) {
            dfs(neighbor, graph, visited, order);
        }
    }
    // Push vertex onto stack after all its neighbors are processed
    order.push(v);
}

int main() {
    int vertices, edges;
    cout << "Enter the number of vertices: ";
    cin >> vertices;
    cout << "Enter the number of edges: ";
    cin >> edges;
    
    // Initialize graph as an adjacency list (0-indexed vertices)
    vector<vector<int>> graph(vertices);
    cout << "Enter each edge (source destination):" << endl;
    for (int i = 0; i < edges; i++) {
        int u, v;
        cin >> u >> v;
        // For a directed graph, add an edge from u to v
        graph[u].push_back(v);
    }
    
    // To keep track of visited vertices
    vector<bool> visited(vertices, false);
    stack<int> order;
    
    // Perform DFS for each vertex that hasn't been visited
    for (int i = 0; i < vertices; i++) {
        if (!visited[i]) {
            dfs(i, graph, visited, order);
        }
    }
    
    // Output the topologically sorted order by popping from the stack
    cout << "Topological Order: ";
    while (!order.empty()) {
        cout << order.top() << " ";
        order.pop();
    }
    cout << endl;
    
    return 0;
}
